From b48a2acebe5eec61b0aa1910ee91c4920c795803 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E2=80=9Chuangyifeng=E2=80=9D?=
 <huangyifeng@eswincomputing.com>
Date: Tue, 4 Jun 2024 14:34:07 +0800
Subject: [PATCH 069/109] WIN2030-15099:fix(feat):Support restart cmd

Changelogs:
	 1.Support restart cmd, when SOM opensbi send restart cmd, do
SOM power off/power on.
          2.when poweroff cmd is triggerd, set timeout time to 6s.

Change-Id: I27129f19d710a3782018caf890964ccb5976166b
---
 Core/Inc/hf_common.h           |  5 ++-
 Core/Src/console.c             |  2 +-
 Core/Src/hf_power_process.c    |  9 ++++-
 Core/Src/hf_protocol_process.c | 62 ++++++++++++++++++++++++++++------
 Core/Src/web-server.c          | 28 ++++++++++++---
 5 files changed, 88 insertions(+), 18 deletions(-)

diff --git a/Core/Inc/hf_common.h b/Core/Inc/hf_common.h
index acc14ee..65c191a 100644
--- a/Core/Inc/hf_common.h
+++ b/Core/Inc/hf_common.h
@@ -51,12 +51,13 @@ typedef enum {
 // Define command types
 typedef enum {
 	CMD_POWER_OFF = 0x01,
-	CMD_RESET,
+	CMD_REBOOT,
 	CMD_READ_BOARD_INFO,
 	CMD_CONTROL_LED,
 	CMD_PVT_INFO,
 	CMD_BOARD_STATUS,
 	CMD_POWER_INFO,
+	CMD_RESTART,    //cold reboot with power off/on
 	// You can continue adding other command types
 } CommandType;
 
@@ -284,6 +285,7 @@ int es_set_som_dip_switch_soft_state_all(int som_dip_switch_soft_ctl_attr, uint8
 int es_eeprom_info_test(void);
 power_switch_t get_som_power_state(void);
 void change_som_power_state(power_switch_t newState);
+void vRestartSOM(void);
 deamon_stats_t get_som_daemon_state(void);
 void change_som_daemon_state(deamon_stats_t newState);
 
@@ -305,6 +307,7 @@ int32_t es_set_rtc_time(struct rtc_time_t *stime);
 int32_t es_get_rtc_date(struct rtc_date_t *sdate);
 int32_t es_get_rtc_time(struct rtc_time_t *stime);
 power_info get_power_info(void);
+int xSOMRestartHandle(void);
 
 #ifdef __cplusplus
 }
diff --git a/Core/Src/console.c b/Core/Src/console.c
index c8d3dab..717d4fc 100644
--- a/Core/Src/console.c
+++ b/Core/Src/console.c
@@ -1100,7 +1100,7 @@ static BaseType_t prvCommandReboot(char *pcWriteBuffer, size_t xWriteBufferLen,
     int ret = HAL_OK;
 
     if (SOM_POWER_ON == get_som_power_state()) {
-        ret = web_cmd_handle(CMD_RESET, NULL, 0, 2000);
+        ret = web_cmd_handle(CMD_REBOOT, NULL, 0, 2000);
         if (HAL_OK != ret) {
             som_reset_control(pdTRUE);
             osDelay(10);
diff --git a/Core/Src/hf_power_process.c b/Core/Src/hf_power_process.c
index 87e9323..7b2f450 100644
--- a/Core/Src/hf_power_process.c
+++ b/Core/Src/hf_power_process.c
@@ -71,7 +71,6 @@ static void atx_power_on(uint8_t turnon);
 static void power_led_on(uint8_t turnon);
 static int pmic_b6out_105v(void);
 power_switch_t get_som_power_state(void);
-void change_som_power_state(power_switch_t newState);
 void set_power_off(void);
 
 // #define MAXTRYCOUNT		10
@@ -597,3 +596,11 @@ void change_som_power_state(power_switch_t newState)
 	taskEXIT_CRITICAL();
 	es_set_som_pwr_last_state(newState);
 }
+
+void vRestartSOM(void)
+{
+	change_som_power_state(SOM_POWER_OFF);
+	/*Delay 2 second*/
+	osDelay(pdMS_TO_TICKS(2000));
+	change_som_power_state(SOM_POWER_ON);
+}
\ No newline at end of file
diff --git a/Core/Src/hf_protocol_process.c b/Core/Src/hf_protocol_process.c
index f5432b5..78d45e1 100644
--- a/Core/Src/hf_protocol_process.c
+++ b/Core/Src/hf_protocol_process.c
@@ -422,6 +422,7 @@ void dump_message(Message data)
 SemaphoreHandle_t xMutex = NULL;
 TimerHandle_t xSomPowerOffTimer;
 TimerHandle_t xSomRebootTimer;
+TimerHandle_t xSomRestartTimer;
 
 // Function to initialize the mutex
 void init_transmit_mutex(void) {
@@ -638,8 +639,13 @@ void handle_notify_mesage(Message *msg)
 {
 	if (CMD_POWER_OFF == msg->cmd_type) {
 		// Here the SOM should at shutdown state in opensbi, we can turn off its power safely
+		vStopSomPowerOffTimer();
+		printf("Poweroff SOM normaly!\n");
 		change_som_power_state(SOM_POWER_OFF);
-		printf("Poweroff SOM normaly, shutdown it now!\n");
+	} else if (CMD_RESTART == msg->cmd_type) {
+		StopSomRestartTimer();
+		printf("Restart SOM normaly!\n");
+		vRestartSOM();
 	}
 }
 
@@ -675,7 +681,7 @@ void handle_som_mesage(Message *msg)
 	}
 }
 
-void vSomPowerOffTimerCallback(TimerHandle_t xSomPowerOffTimer)
+void vSomPowerOffTimerCallback(TimerHandle_t Timer)
 {
 	if (SOM_POWER_OFF != get_som_power_state()) {
 		change_som_power_state(SOM_POWER_OFF);
@@ -686,28 +692,55 @@ void vSomPowerOffTimerCallback(TimerHandle_t xSomPowerOffTimer)
 void TriggerSomPowerOffTimer(void)
 {
 	if (xTimerStart(xSomPowerOffTimer, 0) != pdPASS) {
-		printf("Failed to trigger Som timer!\n");
+		printf("Failed to trigger poweroff Som timer!\n");
 	}
 }
 
-void vSomRebootTimerCallback(TimerHandle_t xSomRebootTimer)
+void vStopSomPowerOffTimer(void)
+{
+	if (xTimerStop(xSomPowerOffTimer, 0) != pdPASS) {
+		printf("Failed to trigger poweroff Som timer!\n");
+	}
+}
+
+void vSomRebootTimerCallback(TimerHandle_t Timer)
 {
 	som_reset_control(pdTRUE);
 	osDelay(10);
 	som_reset_control(pdFALSE);
-	printf("reboot SOM timeout, force reset SOM!\n");
+	printf("Reboot SOM timeout, force reboot SOM!\n");
 }
 
 void TriggerSomRebootTimer(void)
 {
 	if (xTimerStart(xSomRebootTimer, 0) != pdPASS) {
-		printf("Failed to trigger Som timer!\n");
+		printf("Failed to trigger Som reboot timer!\n");
 	}
 }
 void StopSomRebootTimer(void)
 {
 	if (xTimerStop(xSomRebootTimer, 0) != pdPASS) {
-		printf("Failed to trigger Som timer!\n");
+		printf("Failed to trigger Som reboot timer!\n");
+	}
+}
+
+void vSomRestartTimerCallback(TimerHandle_t Timer)
+{
+	printf("Restart SOM timeout, force restart SOM!\n");
+	vRestartSOM();
+}
+
+void TriggerSomRestartTimer(void)
+{
+	if (xTimerStart(xSomRestartTimer, 0) != pdPASS) {
+		printf("Failed to trigger Som restart timer!\n");
+	}
+}
+
+void StopSomRestartTimer(void)
+{
+	if (xTimerStop(xSomRestartTimer, 0) != pdPASS) {
+		printf("Failed to trigger Som restart timer!\n");
 	}
 }
 
@@ -728,8 +761,8 @@ void uart4_protocol_task(void *argument)
 	//Init web server cmd list
 	vListInitialise(&WebCmdList);
 
-	/* Create a timer with a timeout set to 5 seconds */
-	xSomPowerOffTimer = xTimerCreate( "SomPowerOffTimer", pdMS_TO_TICKS(5000),
+	/* Create a timer with a timeout set to 6 seconds */
+	xSomPowerOffTimer = xTimerCreate( "SomPowerOffTimer", pdMS_TO_TICKS(6000),
 			pdFALSE, (void *)0, vSomPowerOffTimerCallback);
 
 	if (xSomPowerOffTimer == NULL) {
@@ -745,6 +778,15 @@ void uart4_protocol_task(void *argument)
 		printf("[%s %d]:Failed to create SOM reboot timer!\n",__func__,__LINE__);
 		return;
 	}
+
+	/* Create a timer with a timeout set to 5 seconds */
+	xSomRestartTimer = xTimerCreate( "SomRestartTimer", pdMS_TO_TICKS(5000),
+			pdFALSE, (void *)0, vSomRestartTimerCallback);
+
+	if (xSomRestartTimer == NULL) {
+		printf("[%s %d]:Failed to create SOM restart timer!\n",__func__,__LINE__);
+		return;
+	}
 	for (;;) {
 		if (xQueueReceive(xUart4MsgQueue, &(msg), portMAX_DELAY)) {
 			if (msg.header == FRAME_HEADER && msg.tail == FRAME_TAIL) {
@@ -764,4 +806,4 @@ void uart4_protocol_task(void *argument)
 			}
 		}
 	}
-}
\ No newline at end of file
+}
diff --git a/Core/Src/web-server.c b/Core/Src/web-server.c
index 7ed703f..be6d5ca 100644
--- a/Core/Src/web-server.c
+++ b/Core/Src/web-server.c
@@ -10443,7 +10443,7 @@ static int change_power_status(int status)
 				ret = HAL_OK;
 				return ret;
 			}
-			// Trigger the Som timer to enusre SOM could poweroff in 5 senconds
+			// Trigger the Som timer to enusre SOM could poweroff anyway
 			TriggerSomPowerOffTimer();
 		} else {
 			web_debug("SOM already power off!\n");
@@ -10466,24 +10466,42 @@ int change_power_lost_resume_attr(int status){//status 0:power off,1:power on
 	return es_set_som_pwr_lost_resume_attr(status);
 }
 
-int reset()
+int xSOMRebootHandle()
 {
 	int ret = HAL_OK;
 
-	ret = web_cmd_handle(CMD_RESET, NULL, 0, 2000);
+	ret = web_cmd_handle(CMD_REBOOT, NULL, 0, 2000);
 	if (HAL_OK != ret) {
 		som_reset_control(pdTRUE);
 		osDelay(10);
 		som_reset_control(pdFALSE);
 		web_debug("Faild to reboot SOM(ret %d), force reset SOM!\n", ret);
 		ret = HAL_OK;
+		return ret;
 	}
-	// Trigger the Som timer to enusre SOM could reboot in 5 senconds
+	// Trigger the Som timer to enusre SOM could reboot anyway
 	TriggerSomRebootTimer();
 	web_debug("web call reset, ret %d\n", ret);
 	return ret;
 }
 
+int xSOMRestartHandle(void)
+{
+	int ret = HAL_OK;
+
+	ret = web_cmd_handle(CMD_RESTART, NULL, 0, 2000);
+	if (HAL_OK != ret) {
+		web_debug("Faild to restart SOM(ret %d), force restart SOM!\n", ret);
+		vRestartSOM();
+		ret = HAL_OK;
+		return ret;
+	}
+	// Trigger the Som timer to enusre SOM could restart anyway
+	TriggerSomRestartTimer();
+	web_debug("call restart, ret %d\n", ret);
+	return ret;
+}
+
 power_info get_power_info(void)
 {
 	power_info power_info = {0};
@@ -11338,7 +11356,7 @@ int get_soc_status()
 
 				}else if(strcmp(path, "/reset")==0 ){
 					web_debug("POST location: reset \n");
-					int reset_ret=reset();
+					int reset_ret=xSOMRebootHandle();
                     char *json_response_patt =NULL;
 					char json_response[BUF_SIZE_64]={0};
 					if(reset_ret==0){
-- 
2.25.1

